server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -242,8 +266,12 @@ public UnassignedInfo(Reason reason, @Nullable String message, @Nullable Excepti
         this.failedAllocations = failedAllocations;
         this.lastAllocationStatus = Objects.requireNonNull(lastAllocationStatus);
         this.failedNodeIds = Collections.unmodifiableSet(failedNodeIds);
-        assert (failedAllocations > 0) == (reason == Reason.ALLOCATION_FAILED) :
-            "failedAllocations: " + failedAllocations + " for reason " + reason;
+        this.lastAllocatedNodeId = lastAllocatedNodeId;
```

Comment:
can we assert that `lastAllocatedNodeIsRestarting == false || lastAllocatedNodeId != null`?

Reply:
I've added such a check.